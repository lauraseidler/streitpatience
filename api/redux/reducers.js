import {
  SET_ONLINE_PLAYERS,
  CREATE_NEW_GAME,
  RECONNECT_PLAYER
} from './action-types';

const DEFAULT_STATE = {
  onlinePlayers: 0,
  games: {}
};

const setOnlinePlayers = (state, action) => ({
  ...state,
  onlinePlayers: action.payload
});

const createNewGame = (state, { payload: { gameId, playerId } }) => ({
  ...state,
  games: {
    ...state.games,
    [gameId]: {
      id: gameId,
      players: [playerId]
    }
  }
});

const reconnectPlayer = (
  state,
  { payload: { gameId, playerId, newPlayerId } }
) => ({
  ...state,
  games: {
    ...state.games,
    [gameId]: {
      ...state.games[gameId],
      players: state.games[gameId].players.map(
        p => (p === playerId ? newPlayerId : p)
      )
    }
  }
});

const rootReducer = (state = DEFAULT_STATE, action) => {
  switch (action.type) {
    case SET_ONLINE_PLAYERS:
      return setOnlinePlayers(state, action);
    case CREATE_NEW_GAME:
      return createNewGame(state, action);
    case RECONNECT_PLAYER:
      return reconnectPlayer(state, action);
    default:
      return state;
  }
};

export default rootReducer;
