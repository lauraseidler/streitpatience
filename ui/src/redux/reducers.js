import { SET_ONLINE_PLAYERS } from './actions';

const DEFAULT_STATE = {
  onlinePlayers: 0
};

const setOnlinePlayers = (state, action) => ({
  ...state,
  onlinePlayers: action.payload
});

const rootReducer = (state = DEFAULT_STATE, action) => {
  switch (action.type) {
    case SET_ONLINE_PLAYERS:
      return setOnlinePlayers(state, action);
    default:
      return state;
  }
};

export default rootReducer;
