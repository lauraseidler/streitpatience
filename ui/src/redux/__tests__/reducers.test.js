import 'jest-localstorage-mock';
import reducers from '../reducers';

test('SET_ONLINE_PLAYERS', () => {
  const state = reducers(
    { onlinePlayers: 0 },
    { type: 'SET_ONLINE_PLAYERS', payload: 1 }
  );

  expect(state).toMatchSnapshot();
});

test('SET_GAME_VIEW', () => {
  const state = reducers(
    { gameView: 'ACTION_BOARD' },
    { type: 'SET_GAME_VIEW', payload: 'ABOUT' }
  );

  expect(state).toMatchSnapshot();
});

test('SET_GAME_VIEW with illegal game view', () => {
  const state = reducers(
    { gameView: 'ACTION_BOARD' },
    { type: 'SET_GAME_VIEW', payload: 'completely bogus game view' }
  );

  expect(state).toMatchSnapshot();
});

test('SET_CURRENT_GAME for first player', () => {
  const state = reducers(
    { gameView: 'NEW_GAME', currentGame: null },
    {
      type: 'SET_CURRENT_GAME',
      payload: { id: '7gg1n2smCT2kkiwaAAAA', players: ['7gg1n2smCT2kkiwaAAAA'] }
    }
  );

  expect(state).toMatchSnapshot();
});

test('PROMPT_RECONNECT', () => {
  const state = reducers(
    { gameView: 'ACTION_BOARD' },
    { type: 'PROMPT_RECONNECT' }
  );

  expect(state).toMatchSnapshot();
});

test('ABORT_RECONNECT', () => {
  const state = reducers(
    { gameView: 'RECONNECT_PROMPT' },
    { type: 'ABORT_RECONNECT' }
  );

  expect(state).toMatchSnapshot();
});
