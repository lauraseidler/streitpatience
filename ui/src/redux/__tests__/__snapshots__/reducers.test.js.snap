// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ABORT_RECONNECT 1`] = `
Object {
  "gameView": "ACTION_BOARD",
}
`;

exports[`PROMPT_RECONNECT 1`] = `
Object {
  "gameView": "RECONNECT_PROMPT",
}
`;

exports[`SET_CURRENT_GAME for first player 1`] = `
Object {
  "currentGame": Object {
    "id": "7gg1n2smCT2kkiwaAAAA",
    "players": Array [
      "7gg1n2smCT2kkiwaAAAA",
    ],
  },
  "gameView": "GAME",
}
`;

exports[`SET_GAME_VIEW 1`] = `
Object {
  "gameView": "ABOUT",
}
`;

exports[`SET_GAME_VIEW with illegal game view 1`] = `
Object {
  "gameView": "ACTION_BOARD",
}
`;

exports[`SET_ONLINE_PLAYERS 1`] = `
Object {
  "onlinePlayers": 1,
}
`;
